api/src/services/import-services/import-critters-service.ts

```diff
@@ -0,0 +1,411 @@
+    const rows = this._getRows(worksheet);
+    const nonStandardColumns = this._getNonStandardColumns(worksheet);
+
+    // Retrieve the dynamic validation config
+    const [validRowTsns, surveyCritterAliases] = await Promise.all([
+      this._getValidTsns(worksheet),
+      this.surveyCritterService.getUniqueSurveyCritterAliases(surveyId)
+    ]);
+    const collectionUnitMap = await this._getCollectionUnitMap(worksheet, validRowTsns);
+
+    // Parse reference data for validation
+    const tsnSet = new Set(validRowTsns.map((tsn) => Number(tsn)));
+    const csvCritterAliases = rows.map((row) => row.animal_id);
+
+    // Track the row validation errors
+    const errors: ValidationError[] = [];
+
+    const csvCritters = rows.map((row, index) => {
+      /**
+       * --------------------------------------------------------------------
+       *                      STANDARD ROW VALIDATION
+       * --------------------------------------------------------------------
+       */
+
+      // SEX is a required property and must be a correct value
+      const invalidSex = !row.sex || !CSV_CRITTER_SEX_OPTIONS.includes(toUpper(row.sex));
+      // WLH_ID must follow regex pattern
+      const invalidWlhId = row.wlh_id && !/^\d{2}-.+/.exec(row.wlh_id);
+      // ITIS_TSN is required and be a valid TSN
+      const invalidTsn = !row.itis_tsn || !tsnSet.has(row.itis_tsn);
+      // ALIAS is required and must not already exist in Survey or CSV
+      const invalidAlias =
+        !row.animal_id ||
+        surveyCritterAliases.has(row.animal_id) ||
+        csvCritterAliases.filter((value) => value === row.animal_id).length > 1;
+
+      if (invalidSex) {
+        errors.push({ row: index, message: `Invalid SEX. Expecting: ${CSV_CRITTER_SEX_OPTIONS.join(', ')}.` });
+      }
+      if (invalidWlhId) {
+        errors.push({ row: index, message: `Invalid WLH_ID. Example format '10-1000R'.` });
+      }
+      if (invalidTsn) {
+        errors.push({ row: index, message: `Invalid ITIS_TSN.` });
+      }
+      if (invalidAlias) {
+        errors.push({ row: index, message: `Invalid ALIAS. Must be unique in Survey and CSV.` });
+      }
+
+      /**
+       * --------------------------------------------------------------------
+       *                      NON-STANDARD ROW VALIDATION
+       * --------------------------------------------------------------------
+       */
+
+      nonStandardColumns.forEach((column) => {
+        const collectionUnitColumn = collectionUnitMap.get(column);
+        // Remove property if undefined or not a collection unit
+        if (!collectionUnitColumn || !row[column]) {
+          delete row[column];
+          return;
+        }
+        // Attempt to find the collection unit with the cell value from the mapping
+        const collectionUnitMatch = collectionUnitColumn.collectionUnits.find((unit) => unit.unit_name === row[column]);
```

Comment:
Maybe by design, but just checking: This is case sensitive, so the cell content has to be cased correctly for this to pass validation.

Reply:
Fixing